lib/base/dependencygraph.cpp

```diff
@@ -32,19 +34,43 @@ void DependencyGraph::RemoveDependency(Object *parent, Object *child)
 		m_Dependencies.erase(child);
 }
 
-std::vector<Object::Ptr> DependencyGraph::GetParents(const Object::Ptr& child)
+std::vector<ConfigObject::Ptr> DependencyGraph::GetParents(const ConfigObject::Ptr& child)
 {
-	std::vector<Object::Ptr> objects;
+	std::vector<ConfigObject::Ptr> objects;
 
-	std::unique_lock<std::mutex> lock(m_Mutex);
+	std::shared_lock<std::shared_mutex> lock(m_Mutex);
 	auto it = m_Dependencies.find(child.get());
 
 	if (it != m_Dependencies.end()) {
-		typedef std::pair<Object *, int> kv_pair;
-		for (const kv_pair& kv : it->second) {
+		for (auto& kv : it->second) {
 			objects.emplace_back(kv.first);
 		}
 	}
 
 	return objects;
 }
+
+std::set<ConfigObject::Ptr> DependencyGraph::GetAllParents(const ConfigObject::Ptr& child)
+{
+	std::set<ConfigObject::Ptr> objects;
+	std::unordered_set<ConfigObject *> toDo {child.get()};
+	std::shared_lock<std::shared_mutex> lock (m_Mutex);
+
+	do {
+		auto current (*toDo.begin());
+
+		toDo.erase(toDo.begin());
+
+		auto it (m_Dependencies.find(current));
+
+		if (it != m_Dependencies.end()) {
+			for (auto& kv : it->second) {
+				if (objects.emplace(kv.first).second) {
+					toDo.emplace(kv.first);
+				}
+			}
+		}
+	} while (!toDo.empty());
```

Comment:
> That attribute would store the combined log severity of that object, i.e. the minimal severity configured on that object or any of its parents. It would be updated when the severity is set, not used

Too bad relationships are runtime malleable, see e.g. dependencies.

> (FYI: I haven't put much time into thinking about whether `DependencyGraph` actually stores a relation that's useful for this use-case or if it might include too much or too little.)

Now as you talk about relationships being "useful for this use-case" I should come out with a thought I had while building this:

This feature doesn't use the DependencyGraph. Instead ConfigObject provides a new virtual method for returning wisely chosen (not checkables across dependencies of course) immediate parents (host for service, checkable for downtime, ...). As it wouldn't include runtime-dynamic parents (checkables across dependencies) and due to #9407, objects could run the algorithm shown above **once on creation** and RAM-cache the result set.

Reply:
> IMAO yes. Especially I want all downtimes and notifications of a service, don't I?

That would also be possible if we said something like "you can only filter on checkables". Each log message would be tagged with the checkable and this would simply be compared in the filter.

If that turns out to be too unflexible, maybe something like you suggested could be an option:

> Instead ConfigObject provides a new virtual method for returning wisely chosen (not checkables across dependencies of course) immediate parents (host for service, checkable for downtime, ...).

But this should be bounded to a low number of objects, maybe like the object itself, a service, and a host at most.

> And maybe even all hostgroup members.

In which case? If you specify a host group in the filter? Or any objects that share a host group?

Reply:
> But this should be bounded to a low number of objects, maybe like the object itself, a service, and a host at most.

Why? For performance reasons per log message? Please ignore them for a moment, I'll handle the performance in general soon.

Instead please have a look at the "parent_affecting_logging" commits from the UX perspective. I.e. user enables verbose logging for X -> Y, Z are affected -> OK?

Reply:
> > But this should be bounded to a low number of objects, maybe like the object itself, a service, and a host at most.
> 
> Why? For performance reasons per log message? Please ignore them for a moment, I'll handle the performance in general soon.

Basically yes, as that would bound the per-message overhead to something like three set lookups.

> Instead please have a look at the "parent_affecting_logging" commits from the UX perspective. I.e. user enables verbose logging for X -> Y, Z are affected -> OK?

I don't really understand the direction this PR is moving towards, but my feeling is that it is becoming more complex than anyone needs or asked for.

Currently, you are just pushing code with little explanation, so please explain in more detail what you're doing and why you're doing that way, including why you think this complexity is necessary.

Reply:
Admittedly you're right, I'm fixing the OP...

Reply:
> > Instead please have a look at the "parent_affecting_logging" commits from the UX perspective. I.e. user enables verbose logging for X -> Y, Z are affected -> OK?
> 
> I don't really understand the direction this PR is moving towards, but my feeling is that it is becoming more complex than anyone needs or asked for.
> 
> Currently, you are just pushing code with little explanation, so please explain in more detail what you're doing and why you're doing that way, including why you think this complexity is necessary.

"necessary" is relative. I mean, I could omit the family tree at all. But then runtime-dynamic things like downtimes and comments wouldn't be auto-included along with their checkable, not to mention scheduled downtimes. Allowing just such edge cases would put up questions like _but why can't I enable a hostgroup and all hosts are affected?_ Now you can. You don't _have to_, but you can.

Just read 1ce044c60 06ea999ba 67fb214ac d98bb316e, but mirror it from the user perspective and imagine some recursion on top. E.g. a hostgroup specified in the logger filter implies also its hosts (06ea999ba), their services (67fb214ac) and the latter's downtimes (d98bb316e).

PS: Yes, I know it doesn't compile yet, but Log#Log() callers are missing anyway.